Improve email confirmation UX

monfresh opened the pull request:
**Why**:
To address the following issues:

- When a user who is already confirmed clicks on the link in their
confirmation email again, they remain on the "Resend confirmation
instructions" screen with an error message that their email has already
been confirmed and they should try signing in. A better UX would be to
redirect them to the appropriate place depending on whether or not they
are signed in.

- If a user confirms, then signs out, and comes back to the site after
a while, forgetting they already confirmed, then clicks "Didn't receive
confirmation instructions?", and resends confirmation instructions to
that same email, they will never receive an email.

**How**:

- If a user is already confirmed and clicks the confirmation link again
while signed out, they are redirected to the sign in page, with the
following flash error message: `Your email address has already been
confirmed. Please sign in.`. If they are already signed in, they are
redirected to the dashboard (which in turn might redirect them to the
part of the account creation process they haven't completed yet).

- If a user is already confirmed and resends confirmation instructions
to their confirmed email, they will receive an email letting them know
their email is already confirmed, with a link to sign in or reset their
password if they forgot it.

- When a user first clicks the confirmation link, the flash notice says
`Your email address has been successfully confirmed. To continue
creating your account, please set a password.` as opposed to just
`Your email address has been successfully confirmed.`. This indicates
that the account is not fully created yet.

- The confirmation email now says `to continue creating your account`,
as opposed to `to finish creating ...` because clicking the confirmation
link does not finish creating the account.

sbc100 commented on app/controllers/users/confirmations_controller.rb:
```diff
@@ -2,6 +2,12 @@ module Users
+    def create
+      super
+
+      UserMailer.already_confirmed(resource).deliver_later if resource.confirmed?
```
I don't know how you (or the rails community) feels about adding comments to code but I think it a comment here would useful.

monfresh replied:
I try to write code in a way that makes it readable and easy to understand without needing comments, but I can certainly add a comment here. In this case, if a resource is already confirmed and tries confirming again, they will receive an email that lets them know they already confirmed. I tried to name the methods in such a way that made that clear. What part did you find confusing?

sbc100 replied:
I guess it wasn't completely clear to me how the user might end up at the #create action, but thats probably because I'm new to rails.

monfresh replied:
No, you're right. I thought about this some more, and it's not immediately clear. I will add a comment. The `create` action is called when a user asks for confirmation instructions to be sent. If a user has already confirmed their email, but asks for instructions again, Devise will not send them an email by default, so we override the `create` method so that an email is sent letting the user know that they've already confirmed.

sbc100 commented on app/controllers/users/confirmations_controller.rb:
```diff
@@ -60,33 +66,44 @@ def process_user_with_password_errors
       set_view_variables
 
       if resource.confirmation_period_expired?
         flash[:error] = UserDecorator.new(resource).confirmation_period_expired_error
```
Should this be flash.now too?

monfresh replied:
Good question. I think we're fine here, but I don't think it hurts to use `flash.now`. I wonder if we should use `flash.now` everywhere because I don't think we have any use cases where we want the flash message to persist after redirection, right?

sbc100 replied:
My understanding is that the flash messages usually are rendered on the page being redirected to.  Its only if you are going to call #render that you would use the #new isn't it?  Maybe I don't understand correctly.

sbc100 replied:
If my understanding is correct I think the majority of our flash messages should not use #now.

monfresh replied:
You are right. In the case below, where I changed the flash to #now, the user clicks on the link in the email, then ends up on ConfirmationsController#show with a flash message saying their email was successfully confirmed, but that they need to set a password to continue creating their account. After a user enters a valid password, they get redirected to the 2FA setup page. If we don't use `#now` for the flash message, it will show up on the 2FA setup page, which is not what we want.

As for the flash error you commented about here, it appears when the user takes too long to click the link in the email. They end up on a screen that asks them to send confirmation instructions again. When they click the button, they get redirected to the home page with a new flash message, so the old one won't get carried over. However, if they decide to navigate away instead of resending instructions, by going to the home page, for example, the message will carry over. Since they can only go to either the home page or the sign up page, perhaps it makes sense to keep this message around in this case to let them know they still need to confirm their email?

sbc100 commented on app/views/user_mailer/already_confirmed.html.slim:
```diff
@@ -0,0 +1,14 @@
+    meta content='text/html; charset=UTF-8' http-equiv='Content-Type'
+  body
+
+    p= 'A request was made to resend confirmation instructions to this email.' \
```
In most other slim templates the style seems to be:

```
p
  'foo bar
  'baz
```

(I think the = is not needed for pure strings.)

monfresh replied:
Good point. We should probably be consistent about this. I'll checks the docs to see what the convention is.

sbc100 commented on spec/features/visitors/sign_up_spec.rb:
```diff
@@ -209,12 +213,12 @@
   end
 
-  scenario 'visitor confirms more than once' do
+  scenario 'visitor confirms more than once while signed in' do
```
Are we not still expecting to see a flash here?

Perhaps the name of this test could be better... something describes what we expect to happen?

`scenario 'signed in users who re-visit the confirmation page are redirected to 2FA setup'`

monfresh replied:
The way I've been using scenarios is to describe the context in the title, and read the expectations in the test to see what is expected to happen. For slow feature tests, it is recommended to combine all expectations in the same test, to avoid running the setup multiple times. For example, instead of something like this, which would run the `before` block for each `it` block:
```ruby
context 'visitor confirms more than once' do
  before do
    create(:user, :signed_up)
    sign_in_user
    send_confirmation_instructions
    click_link_in_email
  end

  it 'does not display flash message' do
    expect(page).to_not have_content('foo')
  end

  it 'redirects to home page' do
    expect(current_url).to eq root_url
  end

  etc...
end
```
You would do this:
```ruby
context 'visitor confirms more than once' do
  it 'redirects to root url without flash messages' do
    create(:user, :signed_up)
    sign_in_user
    send_confirmation_instructions
    click_link_in_email

    expect(page).to_not have_content('foo')
    expect(current_url).to eq root_url
  end
end
```
Sometimes, there might be more expectations than can fit in the `it` title, and I think that's OK. I don't think the `it` title should explain all the expectations.

In general, I prefer the `context` + `it` syntax because it allows you to specify both the scenario and the expectation, whereas with the `scenario` syntax, you have to either describe both in the scenario title, or add comments above the scenario.

Perhaps we should choose one style for consistency?

sbc100 commented on spec/features/visitors/sign_up_spec.rb:
```diff
@@ -301,4 +305,38 @@
+  #   When I click the confirmation link in the email again
+  #   Then I see a message that I've already confirmed
+  #   And I am redirected to the sign in page
+  scenario 'confirmed visitor clicks confirmation link again while signed out' do
```
Followup on earlier comment: Seems that some scenarios describe purely what the user does while others describe the expected action.. do we prefer one of these styles over the other?

monfresh commented:
@sbc100 I think I've addressed your feedback in this commit: https://github.com/18F/identity-idp/pull/182/commits/0bc74641eb45b2f396e8aea63bab7328a5ceeee2

If it looks good, I'll squash.

sbc100 commented on app/views/user_mailer/already_confirmed.html.slim:
```diff
@@ -3,12 +3,15 @@ html
+    p
+      | A request was made to resend confirmation instructions to this email.
+        This email address has already been confirmed. You can sign in with
+        your #{APP_NAME} account by clicking the link below:
```
We should be consistent in the style we use.   At least some of the other slim files use:

```
p
  'Line 1
  'Line 2
```

I don't mind wihch one we choose if they are the equivalent.  Perhaps a followup change to make all the slim files consistent?  I guess in the long run everything will be localised anyway we shouldn't have text in the slim files at all.

monfresh replied:
Good point. I will localize this text, then squash.

sbc100 commented:
lgtm with 1 comment.   (feel free to squash and submit)
